### PR TITLE
Add finalizer to ensure that Runs are stored

### PR DESCRIPTION
**Pull request by khrm**
Added finalizer to TaskRun and PipelineRun to ensure that TaskRun and PipelienRun are stored. This will block deletion until the deadline (Completion + store_deadline) lapses or Runs are stored when done.

# Changes

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you review them:

- [x] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] [Tested your changes locally](https://github.com/tektoncd/results/blob/main/docs/DEVELOPMENT.md) (if this is a code change)
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including functionality, content, code)
- [x] Has a kind label. You can add a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user-facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings)
- [ ] Release notes contain the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
Added finalizer to TaskRun and PipelineRun to ensure that TaskRun and PipelienRun are stored. This will block deletion until the deadline (Completion + store_deadline) lapses or Runs are stored when done.
This finalizer doesn't block if log streaming from watcher or pruner is enabled.
```

### CODE REVIEW

**Review by vdemeester — APPROVED** [review]

File: pkg/watcher/reconciler/pipelinerun/reconciler.go [hunk 1]
```diff
@@ -144,3 +124,66 @@ func isMarkedAsReadyForDeletion(taskRun *pipelinev1.TaskRun) bool {
...
+		storeDeadline = pr.Status.CompletionTime.Add(*r.cfg.StoreDeadline)
+		requeueAfter = storeDeadline.Sub(now)
+		if now.After(storeDeadline) {
+			logging.FromContext(ctx).Errorf("store deadline has passed for pipelinerun %s/%s", pr.Namespace, pr.Name)
+			return nil // Proceed with deletion
+		}
+	}
+
+	if pr.Annotations == nil {
+		logging.FromContext(ctx).Infof("pipelinerun %s/%s annotations are missing, now: %s, storeDeadline: %s, requeueAfter: %s",
```
Comment (khrm) [inline pkg/watcher/reconciler/pipelinerun/reconciler.go:171]: I should probably change these logs for debug.

Comment (khrm) [reply pkg/watcher/reconciler/pipelinerun/reconciler.go]: Changed.

**Review by khrm — COMMENTED** [review]
@enarha I have rebased this. You can review this.

File: pkg/watcher/reconciler/taskrun/reconciler.go [hunk 2]
```diff
@@ -42,43 +39,85 @@ type Reconciler struct {
...
+	}
+
+	// If the completed resource grace period isn't 0, we can't use finalizers
+	// to coordinate deletion because a results pruner is running.
+	if r.cfg.GetCompletedResourceGracePeriod() != 0*time.Second {
+		return nil
 	}
 
-	logger.Infof("Initiating reconciliation for TaskRun '%s/%s'", namespace, name)
+	// Check to make sure the PipelineRun is finished.
```
Comment (enarha) [inline pkg/watcher/reconciler/taskrun/reconciler.go:84]:
```suggestion
	// Check the TaskRun has finished.
```

File: pkg/watcher/reconciler/pipelinerun/reconciler.go [hunk 3]
```diff
@@ -17,32 +17,31 @@ package pipelinerun
...
-	"k8s.io/client-go/tools/cache"
 	"knative.dev/pkg/controller"
 	"knative.dev/pkg/logging"
 	knativereconciler "knative.dev/pkg/reconciler"
 )
 
 // Reconciler represents pipelineRun watcher logic
 type Reconciler struct {
-	// Inline LeaderAwareFuncs to support leader election.
-	knativereconciler.LeaderAwareFuncs
```
Comment (enarha) [inline pkg/watcher/reconciler/pipelinerun/reconciler.go:45]: What's the effect of removing leader awareness? what happens if user is running multiple instances of the watcher? Shouldn't we document the effects of that change in the commit message/release notes?

File: docs/watcher/README.md [hunk 4]
```diff
@@ -63,8 +63,14 @@ metadata:
...
 
 ## Supported version of TaskRun and PipelineRun CR
 
 Results stores PipelineRun and TaskRun as v1. If there are older records, it's possible that they are stored as v1beta1. API server can be configured to start a converter during initialisation.
+
+## Finalizer for blocking deletion
+
+Watcher also implements a finalizer for blocking deletion. This finalizer only blocks when we don't use in-built pruner and streaming logs via watcher.
+
+This requires that we pass `store_deadline` (default 10m). If deletion request comes, then it will block until completion + store_deadline is passed or resource is stored as record.
```
[the remark this answers is not in the document]
Comment (khrm) [reply docs/watcher/README.md:78]: OK. Let me change it to 10m.

[the remark this answers is not in the document]
Comment (khrm) [reply docs/watcher/README.md:78]: Changed.

**Pull request merged**